numba_dpex/core/types/range_types.py

```diff
@@ -0,0 +1,267 @@
+            attr_native = c.unbox(RangeType(typ.ndim), attr_obj)
+            c.pyapi.decref(attr_obj)
+            with cgutils.early_exit_if(c.builder, stack, attr_native.is_error):
+                c.builder.store(cgutils.true_bit, is_error_ptr)
+            ndrange_attr_native_value_map[attr] = attr_native
+
+        global_range_struct = ndrange_attr_native_value_map[
+            "global_range"
+        ].value
+        local_range_struct = ndrange_attr_native_value_map["local_range"].value
+
+        range_datamodel = default_manager.lookup(RangeType(typ.ndim))
+        ndrange_struct.ndim = c.builder.extract_value(
+            global_range_struct,
+            range_datamodel.get_field_position("ndim"),
+        )
+        ndrange_struct.gdim0 = c.builder.extract_value(
+            global_range_struct,
+            range_datamodel.get_field_position("dim0"),
+        )
+        ndrange_struct.gdim1 = c.builder.extract_value(
+            global_range_struct,
+            range_datamodel.get_field_position("dim1"),
+        )
+        ndrange_struct.gdim2 = c.builder.extract_value(
+            global_range_struct,
+            range_datamodel.get_field_position("dim2"),
+        )
+        ndrange_struct.ldim0 = c.builder.extract_value(
+            local_range_struct,
+            range_datamodel.get_field_position("dim0"),
+        )
+        ndrange_struct.ldim1 = c.builder.extract_value(
+            local_range_struct,
+            range_datamodel.get_field_position("dim1"),
+        )
+        ndrange_struct.ldim2 = c.builder.extract_value(
+            local_range_struct,
+            range_datamodel.get_field_position("dim2"),
+        )
+
+    return NativeValue(
+        ndrange_struct._getvalue(), is_error=c.builder.load(is_error_ptr)
+    )
+
+
+@box(RangeType)
+def box_range(typ, val, c):
+    """
+    Convert a native range structure to a Range object.
+    """
+    ret_ptr = cgutils.alloca_once(c.builder, c.pyapi.pyobj)
+    fail_obj = c.pyapi.get_null_object()
+
+    with ExitStack() as stack:
+        range_struct = cgutils.create_struct_proxy(typ)(
+            c.context, c.builder, value=val
+        )
+
+        ndim_obj = c.box(types.int64, range_struct.ndim)
+        with cgutils.early_exit_if_null(c.builder, stack, ndim_obj):
+            c.builder.store(fail_obj, ret_ptr)
+        dim0_obj = c.box(types.int64, range_struct.dim0)
+        with cgutils.early_exit_if_null(c.builder, stack, dim0_obj):
+            c.builder.store(fail_obj, ret_ptr)
+        dim1_obj = c.box(types.int64, range_struct.dim1)
+        with cgutils.early_exit_if_null(c.builder, stack, dim1_obj):
+            c.builder.store(fail_obj, ret_ptr)
+        dim2_obj = c.box(types.int64, range_struct.dim2)
+        with cgutils.early_exit_if_null(c.builder, stack, dim2_obj):
+            c.builder.store(fail_obj, ret_ptr)
+
+        class_obj = c.pyapi.unserialize(c.pyapi.serialize_object(Range))
```

Comment:
Sorry, lack of knowledge, could you explain why do we do this?

Reply:
First a _mea culpa_, I have no idea! Purely "monkey see, monkey do"  from the Numba extension guide.

But, I dug up what it does. Here is the paper trail:

https://github.com/numba/numba/blob/9f74128def1ca51deebfadecc663f0b80f641c57/numba/core/pythonapi.py#L1349

https://github.com/numba/numba/blob/9f74128def1ca51deebfadecc663f0b80f641c57/numba/core/pythonapi.py#L1402

https://github.com/numba/numba/blob/9f74128def1ca51deebfadecc663f0b80f641c57/numba/_helperlib.c#L1058

https://docs.python.org/3/c-api/bytes.html#c.PyBytes_FromStringAndSize

My superficial reading of what is happening here is that for every registered type Numba is maintaining a cache of how to serialize and unserialize an object of that type using the Python `pickle` module. It is a functionality I believe was designed mainly for  on-disk caching of compiled functions and that involves being able to box and unbox argument types. So, we are using that API to call the CPython `PyBytes_FromStringAndSize` to get the `PyObject` structure that represents a `Range` object in this case. Once we have that we are next doing a `PyObject_CallFunctionObjArgs` to call the constructor for that object.

We might need a better CPython expert like @oleksandr-pavlyk for further explanation.